sycl/include/sycl/detail/memcpy.hpp

```diff
@@ -24,7 +24,7 @@ inline void memcpy_no_adl(T1 *Dst, const T2 *Src, size_t Size) {
 #ifdef __SYCL_DEVICE_ONLY__
   __builtin_memcpy(Dst, Src, Size);
 #else
-  std::memcpy(Dst, Src, Size);
+  std::memcpy(static_cast<void *>(Dst), static_cast<const void *>(Src), Size);
```

Comment:
IMO, this should be done on `bit_cast->memcpy_no_adl` call, not in `memcpy_no_adl->memcpy` call.

Also, per https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104855 the warning is correct, so I'm not sure why recent GCCs don't emit it.